### PR TITLE
feat: Windowed Throughput Sampling

### PR DESCRIPTION
**Pull request by yizzlez**
## The Problem

Stripe was originally using the [dynamic throughput sampler](https://github.com/honeycombio/dynsampler-go/blob/main/totalthroughput.go) in order to sample a target throughput. 

There are two tunable parameters with this sampler:
- **T**: target throughput
- **F**: clear frequency

At a high level, the current dynamic throughput sampler operates by adjusting the sampling rate according to **T** in a function called `updateMaps()`. `updateMaps()` runs every **F** seconds and uses the last **F** seconds of data to compute the new rates. The problem with this approach is a fundamental tradeoff that I call the "resolution problem":

- Making **F** smaller makes you more responsive to load spikes, but with the cost of making the sampling decision on less data
- Making **F** larger makes you less responsive to load spikes, but your sample rates will probably be more stable because they are made with more data.

## Solution

_Windowed Throughput Sampling_ allows us to avoid this tradeoff! We introduce two new variables:

- **W**: the number of seconds to lookback
- **U**: the update frequency

Every **U** seconds, we lookback at sample key frequencies W seconds back to recompute our sampling rates. Because we can separately tune **U** and **W**, we don’t have to suffer from the above tradeoffs.

## Implementation

This is much more difficult to implement in a performant way. The naive approach involves keeping track of the timestamp of every trace seen. However, this is `O(# of traces seen)`, which would increase memory usage drastically, especially if traces are being processed at peak volume. 

A more scalable approach involves bucketing. We observe that changes only need to be recomputed every **U** seconds, so we can bundle all events into a single "block". We can keep track of our data structure using a linked list, which makes adding new blocks and dropping expired blocks fast. Much of this processing logic is in `blocklist.go`. 

Other than that, the `updateMaps()` logic matches pretty closely to the original sampler. 

## Results

This new sampler has been in production at Stripe for more than a month now. We've noticed a significant stabilization in our sampling rates: 
Screen Shot 2023-01-09 at 4 44 31 PM

We've also attached continuous profiling to refinery, and observed a pretty low performance overhead penalty for the new sampler:
Screen Shot 2023-02-06 at 4 47 16 PM

### CODE REVIEW

**Comment by kentquirk** [comment]
Hi, @yizzlez. I haven't had a chance yet to look at this more than casually (and you've still got it marked as WIP anyway). I personally will be on holiday until mid-January and will look at it then.

When you're ready, please do include some text on the differences between this and the dynamic and EMA samplers -- in particular, anything on why someone might choose between them.

**Comment by kentquirk** [comment]
Hi, @yizzlez -- this is quite interesting. I will be considering this over the next several days and it appears to be close to something I was about to create myself. I'll keep you posted.

In the meantime, can you please update the title and description in the PR above? The title needs `feat:` at the beginning, and it needs an explanation, please.

**Comment by yizzlez** [comment]
Oops, sorry! I've been procrastinating updating the PR description. Will do that now 😄

**Review by kentquirk — COMMENTED** [review]
This is great stuff. I have a couple of nits, but would love to merge this soon and then include it as a feature in Refinery in an upcoming release. If you've got time to clean those up, I'd appreciate it. 

FWIW, I'm *also* planning to do a throughput version of the EMA sampler that would effectively be this without the blocklist -- it would have a weighting factor instead. I think both could be useful.

File: blocklist.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+func (b *UnboundedBlockList) AggregateCounts(
+	currentIndex int64,
+	lookbackIndex int64,
+) map[string]int {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.doAggregation(currentIndex, lookbackIndex)
+}
+
+// Split out the actual functionality into doAggregation to support better locking semantics.
```
[the remark this answers is not in the document]
Comment (yizzlez) [reply blocklist.go:85]: `doAggregation` is also used later in `BoundedBlockList.AggregateCounts(...)`

File: windowedthroughput.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	return duration.Nanoseconds() / g.DurationPerIndex.Nanoseconds()
+}
+
+func (t *WindowedThroughput) Start() error {
+	// apply defaults
+	if t.UpdateFrequencyDuration == 0 {
+		t.UpdateFrequencyDuration = time.Second
+	}
+	if t.LookbackFrequencyDuration == 0 {
+		t.LookbackFrequencyDuration = 30 * time.Second
```
[the remark this answers is not in the document]
Comment (yizzlez) [reply windowedthroughput.go:89]: Good idea!

**Comment by yizzlez** [comment]
Hi @kentquirk. I've just pushed a new PR addressing your comments. Thanks for the feedback!

I did run the go's race condition tester:

```
$ go test -race blocklist.go blocklist_test.go
ok  	command-line-arguments	6.239s

$ go test -race windowedthroughput.go blocklist.go windowedthroughput_test.go
ok  	command-line-arguments	0.206s
```

**Review by kentquirk — APPROVED** [review]
Excellent work. Thank you!

The plan from here is:
* Merge this to main on Tuesday
* Add an additional function to the Sampler interface to include a count; the existing GetSampleRate will just call that with count=1. (If only we had Python's named arguments...)
* Try to write the EMA Throughput sampler I talked about above -- or decide it's too time-consuming.
* Release it 
* Incorporate the changes into Refinery

We should have a new release of this repo within a couple of weeks. Refinery will take a bit longer. If you need this officially released sooner than that, let me know and I can rearrange the sequence.

**Comment by kentquirk** [comment]
@yizzlez In testing, I have seen the test below fail twice today. I can't tell yet if it's a test bug (where the test itself is assuming more than is true) or an actual code flaw. I'd appreciate any thoughts you might have on it.


```
=== RUN   TestAllConcurrency
    blocklist_test.go:185:
        	Error Trace:	blocklist_test.go:185
        	            				asm_arm64.s:1172
        	Error:      	Not equal:
        	            	expected: map[string]int{"test_key":48}
        	            	actual  : map[string]int{"test_key":49}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (map[string]int) (len=1) {
        	            	- (string) (len=8) "test_key": (int) 48
        	            	+ (string) (len=8) "test_key": (int) 49
        	            	 }
        	Test:       	TestAllConcurrency
```

**Comment by yizzlez** [comment]
Hi @kentquirk, lemme look into this today 🙂

**Pull request merged**